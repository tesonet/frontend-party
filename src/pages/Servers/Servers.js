import React from "react";
import { Redirect } from "react-router-dom";

import ServersList from "../../components/ServersList/ServersList";
import InternalPageHeader from "../../components/InternalPageHeader/InternalPageHeader";

export const Servers = () => {
  const token = localStorage.getItem("token");

  if (!token) {
    return <Redirect to="/" />
  }

  return (
    <div>
      <InternalPageHeader />
      <ServersList />
    </div>
  );
};

export default Servers;