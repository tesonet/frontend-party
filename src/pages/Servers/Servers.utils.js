import { always, cond, equals, ifElse, lt, sort, T } from 'ramda';

export const sortServerList = (sortBy = {}, data = []) =>
  cond([
    [
      equals('name'),
      ifElse(
        () => equals(true, sortBy.isDescending),
        () => sort((a, b) => (lt(a.name, b.name) ? 1 : -1), data),
        () => sort((a, b) => (lt(a.name, b.name) ? -1 : 1), data)
      )
    ],
    [
      equals('distance'),
      ifElse(
        () => equals(true, sortBy.isDescending),
        () => sort((a, b) => b.distance - a.distance, data),
        () => sort((a, b) => a.distance - b.distance, data)
      )
    ],
    [T, always(data)]
  ])(sortBy.key);
