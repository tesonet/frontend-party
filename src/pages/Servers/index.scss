@import "../../variables.scss";

.servers {
  &__page-header {
    padding: 40px 20px;

    > div {
      display: flex;
      align-items: center;
      justify-content: space-between;
    }

    i {
      padding-right: 5px;
    }
  }

  &__logo {
    height: 30px;
  }
  
  &__list-header {
    padding: 25px 20px;
    line-height: $font-size;
    border: 1px solid $color-gray;
    color: $color-gray-dark;
    border-bottom: none;
    background-color: $color-gray-light;

    > div {
      display: flex;
      justify-content: space-between;
    }
  }

  &__list-item {
    padding: 25px 20px;
    line-height: $font-size;
    border: 1px solid $color-gray;

    & + & {
      border-top: none;
    }

    > div {
      display: flex;
      justify-content: space-between;
    }
  }

}