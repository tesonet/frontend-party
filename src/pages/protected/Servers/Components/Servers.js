import React, {Component} from 'react';
import ServerListItem from './ServerListItem'

export default class Servers extends Component {

    componentDidMount() {
        if (!this.state.servers.length) {
            this.props.fetchList()
        }
    }

    static getDerivedStateFromProps(nextProps) {
        return {
            servers: nextProps.servers || []
        }
    }

    state = {
        servers: [],
        serversSortedAscending: true,
        distanceSortedAscending: false
    }

    sortByName = () => {
        this.state.serversSortedAscending ?
            this.setState({
                servers: this.state.servers.sort((a, b) => {
                    if (a.name > b.name)
                        return -1
                    if (a.name < b.name)
                        return 1
                    return 0
                }),
                serversSortedAscending: !this.state.serversSortedAscending
            }) :
            this.setState({
                servers: this.state.servers.sort((a, b) => {
                    if (a.name < b.name)
                        return -1
                    if (a.name > b.name)
                        return 1
                    return 0
                }),
                serversSortedAscending: !this.state.serversSortedAscending
            })
    }

    sortByDistance = () => {
        this.state.distanceSortedAscending ?
            this.setState({
                servers: this.state.servers.sort((a, b) =>
                    a.distance - b.distance || a.distance - b.distance),

            })
            :
            this.setState({
                servers: this.state.servers.sort((a, b) =>
                    a.distance - b.distance || a.distance - b.distance)
            })
    }

    render() {
        console.log(this.state);
        return (
            <div className="inner-page">
                <div className="list">
                    <div className="list_header">
                        <button className="btn" onClick={this.sortByName}>SERVER</button>
                        <button className="btn" onClick={this.sortByDistance}>DISTANCE</button>
                    </div>
                    <ul className="list_items">
                        {this.state.servers.map((server, index) => (
                            <ServerListItem
                                key={index}
                                server={server}
                                addToFavorites={this.props.addToFavorites}
                            />
                        ))}
                    </ul>
                </div>
            </div>
        );
    }
}
