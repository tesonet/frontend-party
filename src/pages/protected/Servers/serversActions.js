import axios from 'axios'

export const FETCH_LIST_START = '[ASYNC]FETCH_LIST_START'
export const FETCH_LIST_SUCCESS = '[ASYNC]FETCH_LIST_SUCCESS'
export const FETCH_LIST_ERROR = '[ASYNC]FETCH_LIST_ERROR'

export const ASINC_ACTION_START = '[ASYNC]ASINC_ACTION_START'
export const ASINC_ACTION_END = '[ASYNC]ASINC_ACTION_END'


export const ADD_TO_FAVORITES = '[ADD_TO_FAVORITES'
export const REMOVE_FROM_FAVORITES = 'REMOVE_FROM_FAVORITES'

const listUrl = 'http://playground.tesonet.lt/v1/servers'

const fetchList = () => (
    async (dispatch) => {
        try {
            const config = {
                method: 'get',
                url: listUrl,
                headers: {'Authorization': localStorage.getItem('token')}
            }

            const response = await axios(config)
            dispatch({
                type: FETCH_LIST_SUCCESS,
                payload: response
            })

        } catch (e) {
            dispatch({
                type: FETCH_LIST_ERROR,
                error: e
            })
        }
    }
)

const addToFavorites = (server) => ({
    type: ADD_TO_FAVORITES,
    server
})

const removeFromFavorites = (id) => ({
    type: REMOVE_FROM_FAVORITES,
    id: id
})

export {addToFavorites, removeFromFavorites, fetchList}
