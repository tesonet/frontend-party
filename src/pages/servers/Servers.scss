//Mixins

@mixin space-between {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    padding: 15px;
}

//Variables

$logout-btn-hv: #99cc33;
$border-color: #e6e6e6;
$list-header-bg: #f5f5f5;
$list-text-color: #666666;
$header-text-color: #999999;

.servers-page {
    .page-header-container {
        @include space-between;
    	align-items:center;
        padding: 40px 15px;

        .logout-button {
            position: relative;
            padding-left: 25px;
            cursor: pointer;
            font-size: 14px;

            &:hover {
                color: $logout-btn-hv;
            }

            &:before {
                content: '';
                position: absolute;
                left: 0;
                top: 3px;
                width: 15px;
                height: 15px;
                background: url('../../assets/icons/logout.png');
                background-repeat: no-repeat;
                background-size: cover;
            }
        }
    }

    .list-header-container {
        @include space-between;
        color: $header-text-color;
        background-color: $list-header-bg;
        border-bottom: 1px solid $border-color;
        border-top: 1px solid $border-color;
        div {
        	&:hover {
        		opacity: 0.6;
        	}
        	cursor: pointer;
        }
    }

    .servers-list-row {
    	color: $list-text-color;
        @include space-between;
        border-bottom: 1px solid $border-color;
    }
}