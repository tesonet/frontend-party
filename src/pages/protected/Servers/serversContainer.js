import { connect } from 'react-redux'
import uuid from 'uuid'
import ServerList from './Components/ServerList'
import { fetchList, addToFavourites, sortByDistance, sortByName, removeFromFavourites } from './serversActions'

const mapDispatchToProps = dispatch => ({
    fetchList: (token) => dispatch(fetchList(token)),
    addToFavourites: server => dispatch(addToFavourites(server)),
    sortByDistance: () => dispatch(sortByDistance()),
    sortByName: () => dispatch(sortByName()),
    removeFromFavourites: server => dispatch(removeFromFavourites(server))
})

export const mapStateToProps = state => ({
    servers: addId(state.serversReducer.servers)
})

const addId = countries => countries.map(country => Object.assign({}, country, {
    id: uuid.v1()
}))

export default connect(mapStateToProps, mapDispatchToProps)(ServerList)

