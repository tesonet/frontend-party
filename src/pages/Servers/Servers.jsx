import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { Page } from 'containers';
import { Table } from 'components/PageElements';

import { actions as serversActions } from 'store/servers';

import { sortServerList } from './Servers.utils';

const ServerList = () => {
  const dispatch = useDispatch();
  const isLoading = useSelector(state => state.servers.isLoading);
  const error = useSelector(state => state.servers.error);
  const servers = useSelector(state => state.servers.servers);

  useEffect(() => {
    dispatch(serversActions.getServers());
  }, []);

  return (
    <Page withDashboard isLoading={isLoading}>
      {error ? (
        error
      ) : (
        <Table
          headerItems={[
            { name: 'SERVER', key: 'name' },
            { name: 'DISTANCE', key: 'distance' }
          ]}
          items={servers.map(item => ({
            key: `${item.name}-${item.distance}`,
            ...item
          }))}
          sortingFunction={sortServerList}
        />
      )}
    </Page>
  );
};

export default ServerList;
