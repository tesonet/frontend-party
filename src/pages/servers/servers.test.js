import React from 'react';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import { createMockStore } from 'redux-test-utils';
import Servers from './index';

const serverArray = [
  {
    name: 'aa',
    distance: 34,
  },
  {
    name: 'aaa',
    distance: 4,
  },
];
const store = createMockStore({
  servers: {
    servers: serverArray,
  },
});

it('Servers should match snapshot', () => {
  const wrapper = shallow(
    <Provider store={store}>
      <Servers
        getServers={() => {}}
        servers={serverArray}
      />
    </Provider>,
  );
  expect(wrapper).toMatchSnapshot();
});
