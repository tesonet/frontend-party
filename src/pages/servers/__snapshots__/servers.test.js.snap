// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Servers should match snapshot 1`] = `ShallowWrapper {}`;
